[Law Wei Jie] iP

bakano98 opened the pull request:
## Burp (modified Duke)
> "Probably a better looking Duke" - me

Burp helps you to keep track of things that you need to do. Burp is super simple! It is:
- Text-based
- Easy to learn (simple commands!)
- ~~Will possibly complete your task for you~~
- Super easy to use

Best of all? Getting Burp is super simple as well:
1. Download it [here](https://github.com/bakano98/ip/releases/tag/A-Jar)
2. Run using `java -jar _filename._jar`
3. Add your tasks
4. Keep track of your tasks easily

_Oh, and did I mention?_ **It's completely free!**

Features of Burp:
- [x] Track your to-dos
- [x] Track your deadlines
- [x] Track your events
- [ ] GUI (coming soon!)
- [ ] Better looking UI (after GUI)

### Commands to use Burp:
1. `todo (todo-descriptor)`
2. `event (event-descriptor) /at (event-date/time)`
3. `deadline (deadline-descriptor) /by (deadline)`
4. `list` (to list out all items in your to-do-list)
5. `find (search-term)` (to find certain items in your to-do-list)
6. `delete (task-number)` (to delete items from your to-do-list)
7. `mark (task-number)` (to mark an item as done)
8. `unmark (task-number)` (to unmark an item as not done)
9. `bye` (to exit)

## 
If you are a Java Programmer and you are interested, here is the driver code:
```
    public void run() throws DukeException, IOException {
        // Declaration of variables
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        String cmd;

        // Print out the welcome message and await user input
        Ui.showWelcome();
        while (!(cmd = br.readLine()).equals("bye")) {
            String commandType = cmd.split(" ")[0];
            ui.burpReply(ui.determineType(commandType), toDoList, cmd, this.storage);
        }
        Ui.showBye();
    }
```

Commit:
Add Gradle support

bakano98 committed:
Level-1 : Greet, Echo, Exit

bakano98 committed:
Level-2: Add, List

bakano98 committed:
create Task

bakano98 committed:
Level-3: Mark as Done

also added some error handling

bakano98 committed:
Level-4: ToDo, Event, Deadline

bakano98 committed:
Level-5: Handle Errors

bakano98 committed:
Level-6: Delete

bakano98 committed:
A-TextUiTesting

bakano98 committed:
A-Enums

use enums to determine replies from Burp

bakano98 committed:
Level-7: Save

bakano98 committed:
Update .gitignore

bakano98 committed:
Update Duke.java

put writing logic into a function

bakano98 committed:
merge branch-Level-7

put writing logic into a function

bakano98 committed:
Merge pull request #1 from bakano98/branch-Level-7

merge branch-Level-7: Save

bakano98 committed:
Level-8: Dates and Times

bakano98 committed:
Merge pull request #2 from bakano98/branch-Level-8

merge branch-Level-8

bakano98 committed:
styling

follow CS2103T switch-case styling

bakano98 committed:
update details

bakano98 committed:
A-MoreOOP

bakano98 committed:
A-Packages

bakano98 committed:
A-Packages

rename packages to standard naming convention

bakano98 committed:
Update Parser.java

bakano98 committed:
update commands

forgot to extend Command

bakano98 committed:
change return type

parseCommands now returns a Command. for easier time with JUnit testing

bakano98 committed:
update to throw DukeException

bakano98 committed:
A-JUnit

bakano98 committed:
A-TextUiTesting

bakano98 committed:
Update Deadline.java

bakano98 committed:
A-TextUiTesting

update text

bakano98 committed:
update save logic

save task on every add

bakano98 committed:
Update imports

Follow coding standard

bakano98 committed:
Update runtest.bat

bakano98 committed:
A-JavaDoc

Add JavaDoc comments for all files

bakano98 committed:
Update .gitignore

Add manifest file

bakano98 committed:
A-CodingStandard

Change order of import statements.

Import statements are ordered to adhere to the Java Coding Standard as specified in the module website.

bakano98 committed:
A-CodingStandard

Change storage to private to adhere to coding standards.

As a result, storage has to be passed down to the following:
Ui.burpReply(...), Parser.parseCommand(...), and the JUnit tests

bakano98 committed:
A-CodingStandard

Separate logical unit blocks by blank lines

bakano98 committed:
Level-9

Add FindCommand and Ui.Reply.FIND

Ui.Reply.FIND is an identifier to determine if the user input is a FindCommand.

FindCommand will then run and find according to a given search term.

Other updates are just styling updates.

bakano98 committed:
Merge pull request #3 from bakano98/branch-Level-9

Level-9

bakano98 committed:
Merge pull request #4 from bakano98/add-gradle-support

Add Gradle support

bakano98 committed:
A-Gradle

Set up gradle to run unit tests

bakano98 committed:
A-CheckStyle

Use CheckStyle and updated according to CheckStyle feedback

bakano98 committed:
Merge pull request #5 from bakano98/branch-A-CheckStyle

A-CheckStyle

bakano98 committed:
Update Duke.java

Set default filepath as a hardcoded value

bakano98 committed:
Basic GUI

Add "Hello World" as GUI to kick-off the creation of GUI for Duke

bakano98 committed:
Level-10: GUI

Update files to output message in GUI

Add new methods to get response from Duke to GUI

Add JavaDoc comments

bakano98 committed:
Add ByeCommand

Add ByeCommand to gracefully handle a "bye". This causes the GUI to close instantly.

bakano98 committed:
Update GUI stuff

Remove text-based UI printing

Add logic to handle a "bye"

Update UI to look better

bakano98 committed:
Reformat, add bg colour

Reformat Duke's responses

Add background colour to Duke's response and user's inputs

bakano98 committed:
Fix checkstyle errors

bakano98 committed:
Level-10: GUi with FXML

Use FXML to do GUi stuff

bakano98 committed:
Update GUi

Move position of textboxes

bakano98 committed:
Update error messages

Remove emoji from error messages

bakano98 committed:
Improve GUI

Add VBoxes for styling to center components properly

bakano98 committed:
Update styling

Update styling to follow CheckStyle rules

bakano98 committed:
Merge branch 'branch-Level-10'

bakano98 committed:
Add title

Add title for the launched application.

bakano98 committed:
Add custom images

Change image for Duke and User

jinnhl commented on src/main/java/duke/main/Ui.java:
```diff
@@ -0,0 +1,111 @@
+     */
+    public static void showBye() {
+        Ui.setDukeResponse(Parser.formatMsg("Bye. Hope to see you again soon!"));
+    }
```
I like that the method names are consistently very descriptive in what it does.

jinnhl commented on src/main/java/duke/main/Parser.java:
```diff
@@ -0,0 +1,69 @@
+     * @return Returns a Command to indicate what the last Command run was
+     * @throws DukeException when a WrongCommand is given
+     */
+    public static Command parseCommands(Ui.Reply type, TaskList toDoList,
+                                        String cmd, Storage storage) throws DukeException {
+        String[] cmdSplit = cmd.split(" ");
+        switch (type) {
+        case LIST:
+            return new ListCommand(toDoList, cmd);
+        case TODO:
+            return new AddToDoCommand(toDoList, cmd, storage);
+        case DEADLINE:
+            return new AddDeadlineCommand(toDoList, cmd, storage);
+        case EVENT:
+            return new AddEventCommand(toDoList, cmd, storage);
+        case MARK:
+            return new MarkCommand(toDoList, Integer.parseInt(cmdSplit[1]) - 1);
+        case UNMARK:
+            return new UnmarkCommand(toDoList, Integer.parseInt(cmdSplit[1]) - 1);
+        case DELETE:
+            return new DeleteCommand(toDoList, Integer.parseInt(cmdSplit[1]) - 1);
+        case FIND:
+            return new FindCommand(toDoList, cmd);
+        case BYE:
+            return new ByeCommand();
+        default:
+            return new WrongCommand();
```
Example of good implementation of "Make the Happy Path Prominent"

Fenway17 replied:
I like that this part is very clean and simple.

jinnhl commented on src/main/java/duke/commands/UnmarkCommand.java:
```diff
@@ -0,0 +1,47 @@
+        try {
+            // Unmark the task
+            Task taskToUnmark = toDoList.get(numberToUnmark);
+            taskToUnmark.unmark();
```
I like the separation of the method unmark() by declaring the task to be unmarked as taskToUnmark first. This maximises readability

jinnhl commented on src/main/java/duke/commands/ListCommand.java:
```diff
@@ -0,0 +1,37 @@
+        // Prints out the list of items in the TaskList
+        toPrint += "(Total " + toDoList.size() + ") Here's what you have:\n";
+        for (int i = 0; i < toDoList.size(); i++) {
+            toPrint += "" + (i + 1) + "." + toDoList.get(i).toString() + "\n";
```
Would it be better for readability if we declutter the code between line 31-33 by declaring toDoList.size() as a variable first before using it? This would reduce the occurrence of toDoList from 3 to 1

jinnhl reviewed (commented):
Overall the code quality is excellent, such as the use of descriptive methods, very well done comments and separation of code that helps the readability tremendously. Minor optimization needed. Good Job!

Fenway17 commented on src/main/java/duke/commands/AddDeadlineCommand.java:
```diff
@@ -0,0 +1,65 @@
+import duke.tasks.Deadline;
+
+
+/**
+ * AddDeadlineCommand is a Command.
+ * This Command is used to add Deadlines to the TaskList.
+ */
+public class AddDeadlineCommand extends Command<String> {
```
Perhaps it might be better to rename AddDeadlineCommand to just DeadlineCommand?

Fenway17 replied:
Using verbs as class names might get confusing when calling constructors and methods

Fenway17 commented on src/main/java/duke/commands/AddDeadlineCommand.java:
```diff
@@ -0,0 +1,65 @@
+     */
+    public AddDeadlineCommand(TaskList toDoList, String cmd, Storage storage) throws DukeException {
+        this.storage = storage;
+        this.runCommand(toDoList, cmd);
```
Calling a method works for this, but not conventional to call methods inside constructors as the method calls can not be seen in code calling the constructor and might get confusing. I suggest calling the constructor and then calling runCommand in the code manually after the constructor.

Fenway17 commented on src/main/java/duke/commands/AddDeadlineCommand.java:
```diff
@@ -0,0 +1,65 @@
+     * @param cmd      the user input to Burp
+     * @throws DukeException if no Task description is given
+     */
+    public void runCommand(TaskList toDoList, String cmd) throws DukeException {
+        try {
+            // Reformat the command to get the Task's description and deadline
+            String[] deadlineDetails = cmd.split("deadline")[1].split("/by");
+            String deadlineName = deadlineDetails[0];
+            String deadline = deadlineDetails[1];
+
+            // Create a new Deadline to add to TaskList
+            Deadline newDeadline = new Deadline(deadlineName, false, deadline);
+            toDoList.add(newDeadline);
+
+            // Print out the formatted message after adding to TaskList
+            Ui.setDukeResponse(Parser.formatMsg("Got it. I've added this task:\n" + newDeadline
+                    + "\nNow you have " + toDoList.size() + " tasks in the list."));
+
+            // Write the contents of the TaskList to our storage
+            storage.writeFileContent(toDoList);
+        } catch (IndexOutOfBoundsException e) {
+            Ui.setDukeResponseError(Parser.formatMsg("OOPS!!! The description of a deadline cannot be empty."));
+            throw new DukeException(Parser.formatMsg("OOPS!!! The description of a deadline cannot be empty."));
+        } catch (IOException e) {
+            throw new DukeException(Parser.formatMsg("IOException caught") + e);
+        }
+    }
+}
```
I like the descriptive comments for different parts of the method!

Fenway17 commented on src/main/java/duke/main/Duke.java:
```diff
@@ -0,0 +1,54 @@
+    private Ui ui;
+
+    private final String filepath = "./tasklist.txt";
+    private boolean isFirstStartup = true;
```
Is this boolean unused?

Fenway17 reviewed (commented):
Overall code is clean and easy to read through, with only a few non-conventional implementations that might be confusing in a different context/project.

bakano98 committed:
A-Assertions

Add assertions at certain points of the code.

Assertions are used to ensure correctness of the code at certain points

One of the few important points are creation of new Commands. Ensure that new Commands are of the correct, specific type (not just a Command).

As a step to ensure correctness in the code base, let's add assertions at these important points. Doing so will ensure that code correctness is maintained as much as possible before new commits.

bakano98 committed:
Merge pull request #6 from bakano98/A-Assertions

A-Assertions

bakano98 committed:
A-CodeQuality

Improve code quality by tidying up unuseful comments/commands.

There are several ways to improving code quality. One of these is readability.

Removing obvious comments, unused variables is one of the ways to improve code quality.

Ensure that comments added are useful in order to improve readability. Furthermore, remove unused variables.

bakano98 committed:
Merge pull request #7 from bakano98/A-CodeQuality

A-CodeQuality

bakano98 committed:
Remove unused imports

bakano98 committed:
Enable assertions in gradle

bakano98 committed:
Change JavaFX API version

Change JavaFX API version to match with runtime version

bakano98 committed:
Fix some formatting stuff

bakano98 committed:
C-DetectDuplicates

Add a DuplicateCommand class and logic to identify when a duplicate Task is about to be added.

Logically, duplicate tasks should not be added to the list. Therefore, it is necessary to handle the case where a duplicate item is detected.

A duplicate item will not be added to the TaskList. Instead, a message (GUI) will be shown to the user to let them know they are adding a duplicate task.

This can be even more comprehensive by detecting if there are conflicting date/times. Doing so will ensure the user is more aware of their schedule.

bakano98 committed:
Add additional test to junit

For testing DuplicateCommand

bakano98 committed:
Merge pull request #8 from bakano98/branch-addExtensions

Branch add extensions

bakano98 committed:
Create gradle.yml

bakano98 committed:
Update JavaDocs

bakano98 committed:
Add readme and Ui showcase image

bakano98 committed:
Update Deadline.java

bakano98 committed:
Set theme jekyll-theme-slate

bakano98 committed:
Update build.gradle

bakano98 committed:
Merge branch 'master' of https://github.com/bakano98/ip

bakano98 committed:
Update build.gradle

Fix for generating jar files to run properly in Linux, MacOS
